File: app/src/main/java/com/kakacat/minitool/common/util/SystemUtil.java
package com.kakacat.minitool.common.util;

import android.app.Activity;
import android.app.Service;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.BatteryManager;
import android.os.Vibrator;
import android.util.Log;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.DataOutputStream;
import java.io.File;
import java.io.IOException;
import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.NetworkInterface;
import java.net.SocketException;
import java.util.Enumeration;
import java.util.Objects;

public class SystemUtil {

    public static void log(String log) {
        Log.d("hhh", log);
    }

    public static void modifyDpi(String val) {
        String[] commands = new String[]{
                "wm density " + val + "\n"
        };
        executeLinuxCommand(commands, true, false);
    }

    public static void executeLinuxCommand(String[] commands, boolean needRoot, boolean waitFor) {
       try {
            if (needRoot) {
                Process process = Runtime.getRuntime().exec("su");
                DataOutputStream os = new DataOutputStream(process.getOutputStream());
                for (String command : commands) {
                    os.writeBytes(command);
                }
                os.flush();
                if (waitFor) {
                    process.waitFor();
                }
            } else {
                Runtime runtime = Runtime.getRuntime();
                for (String cmd : commands) {
                    runtime.exec(new String[]{"/bin/sh", "-c", cmd});
                }
            }
        } catch (IOException | InterruptedException e) {
            e.printStackTrace();
        }
    }

    public static void createFile(@NotNull File file, boolean needRoot){
        String[] commands;
        if(file.isDirectory()){
            commands = new String[]{
                    "mkdir " + file.getAbsolutePath() + "\n"
            };
            executeLinuxCommand(commands,needRoot,true);
        }else{
            File parentFile = file.getParentFile();
            assert parentFile != null;
            commands = new String[]{
                    "mkdir " + parentFile.getAbsolutePath() + "\n",
            };
            executeLinuxCommand(commands,needRoot,true);
            try {
                file.createNewFile();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public static CharSequence getDataFormClipBoard(@NotNull Context context) {
        ClipboardManager clipboardManager = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
        return Objects.requireNonNull(clipboardManager.getPrimaryClip()).getItemAt(0).getText();
    }

    public static void copyToClipboard(@NotNull Context context, String label, CharSequence content) {
        ClipboardManager cm = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
        cm.setPrimaryClip(ClipData.newPlainText(label, content));
    }

    public static int getElectricity(@NotNull Context context) {
        BatteryManager batterymanager = (BatteryManager) context.getSystemService(Context.BATTERY_SERVICE);
        batterymanager.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY);
        return batterymanager.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY);
    }

    public static void setBatteryLevel(String val) {
        String[] commands = new String[]{
                "dumpsys battery set level " + val + "\n"
        };
        executeLinuxCommand(commands, true, false);
    }

    public static void resetBattery() {
        String[] commands = new String[]{
                "dumpsys battery reset" + "\n"
        };
        executeLinuxCommand(commands, true, false);
    }

    public static void vibrate(@NotNull Context context, long milliseconds) {
        Vibrator vibrator = (Vibrator) context.getSystemService(Service.VIBRATOR_SERVICE);
        vibrator.vibrate(milliseconds);
    }

    public static void openMarket(@NotNull Context context) {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setData(Uri.parse("market://details?id=" + context.getPackageName()));
        context.startActivity(intent);
    }

    public static void openAppDetailInSetting(@NotNull Activity activity) {
        Intent intent = new Intent();
        intent.setAction("android.settings.APPLICATION_DETAILS_SETTINGS");
        intent.setData(Uri.parse("package:" + activity.getPackageName()));
        activity.startActivity(intent);
    }

    public static void openAppDetailInSetting(@NotNull Activity activity,String packageName) {
        Intent intent = new Intent();
        intent.setAction("android.settings.APPLICATION_DETAILS_SETTINGS");
        intent.setData(Uri.parse("package:" + packageName));
        activity.startActivity(intent);
    }

    @Nullable
    public static String getLocalIPAddress() {
        try {
            for (Enumeration<NetworkInterface> mEnumeration = NetworkInterface.getNetworkInterfaces(); mEnumeration.hasMoreElements(); ) {
                NetworkInterface intf = mEnumeration.nextElement();
                for (Enumeration<InetAddress> enumIPAddr = intf.getInetAddresses(); enumIPAddr.hasMoreElements(); ) {
                    InetAddress inetAddress = enumIPAddr.nextElement();
                    if (inetAddress instanceof Inet4Address && !inetAddress.isLoopbackAddress()) {
                        return inetAddress.getHostAddress();
                    }
                }
            }
        } catch (SocketException e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: app/src/main/java/com/kakacat/minitool/common/util/SystemUtil.kt
package com.kakacat.minitool.common.util

import android.app.Activity
import android.app.Service
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.BatteryManager
import android.os.Vibrator
import android.util.Log
import java.io.DataOutputStream
import java.io.File
import java.io.IOException
import java.net.Inet4Address
import java.net.NetworkInterface
import java.net.SocketException

object SystemUtil {

    @JvmStatic
    fun log(log: String = "hello world") {
        Log.d("hhh", log)
    }

    @JvmStatic
    fun modifyDpi(`val`: String) {
        val commands = arrayOf(
                "wm density $`val`\n"
        )
        executeLinuxCommand(commands, true)
    }

    @JvmStatic
    fun executeLinuxCommand(commands: Array<String>, needRoot: Boolean = false, waitFor: Boolean = false) {
        try {
            if (needRoot) {
                val process = Runtime.getRuntime().exec("su")
                val os = DataOutputStream(process.outputStream)
                for (command in commands) {
                    os.writeBytes(command)
                }
                os.flush()
                if (waitFor) {
                    process.waitFor()
                }
            } else {
                val runtime = Runtime.getRuntime()
                for (cmd in commands) {
                    runtime.exec(arrayOf("/bin/sh", "-c", cmd))
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    fun createFile(file: File, needRoot: Boolean = false,waitFor: Boolean = true) {
        val commands: Array<String>
        if (file.isDirectory) {
            commands = arrayOf("mkdir ${file.absolutePath}")
            executeLinuxCommand(commands, needRoot, waitFor)
        } else {
            val parentFile = file.parentFile!!
            commands = arrayOf("mkdir ${parentFile.absolutePath}")
            executeLinuxCommand(commands, needRoot, waitFor)
            file.createNewFile()
        }
    }

    @JvmStatic
    fun getDataFormClipBoard(context: Context): CharSequence {
        val clipboardManager = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        return clipboardManager.primaryClip!!.getItemAt(0).text
    }

    @JvmStatic
    fun copyToClipboard(context: Context, label: String?, content: CharSequence?) {
        val cm = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        cm.setPrimaryClip(ClipData.newPlainText(label, content))
    }

    @JvmStatic
    fun getElectricity(context: Context): Int {
        val batteryManager = context.getSystemService(Context.BATTERY_SERVICE) as BatteryManager
        batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)
        return batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)
    }

    @JvmStatic
    fun setBatteryLevel(`val`: String) {
        val commands = arrayOf(
                "dumpsys battery set level $`val`\n"
        )
        executeLinuxCommand(commands, true)
    }

    @JvmStatic
    fun resetBattery() {
        val commands = arrayOf(
                """
                    dumpsys battery reset

                    """.trimIndent()
        )
        executeLinuxCommand(commands, true)
    }

    fun vibrate(context: Context, milliseconds: Long) {
        val vibrator = context.getSystemService(Service.VIBRATOR_SERVICE) as Vibrator
        vibrator.vibrate(milliseconds)
    }

    @JvmStatic
    fun openMarket(context: Context) {
        val intent = Intent(Intent.ACTION_VIEW)
        intent.data = Uri.parse("market://details?id=" + context.packageName)
        context.startActivity(intent)
    }

    fun openAppDetailInSetting(context: Context) {
        val intent = Intent()
        intent.action = "android.settings.APPLICATION_DETAILS_SETTINGS"
        intent.data = Uri.parse("package:" + context.packageName)
        context.startActivity(intent)
    }

    fun openAppDetailInSetting(activity: Activity, packageName: String) {
        val intent = Intent()
        intent.action = "android.settings.APPLICATION_DETAILS_SETTINGS"
        intent.data = Uri.parse("package:$packageName")
        activity.startActivity(intent)
    }


    val localIPAddress: String?
        get() {
            try {
                val mEnumeration = NetworkInterface.getNetworkInterfaces()
                while (mEnumeration.hasMoreElements()) {
                    val inf = mEnumeration.nextElement()
                    val enumIPAddress = inf.inetAddresses
                    while (enumIPAddress.hasMoreElements()) {
                        val inertAddress = enumIPAddress.nextElement()
                        if (inertAddress is Inet4Address && !inertAddress.isLoopbackAddress()) {
                            return inertAddress.getHostAddress()
                        }
                    }
                }
            } catch (e: SocketException) {
                e.printStackTrace()
            }
            return null
        }
}